src/contrib/cluster/Akka.Cluster.Sharding/ShardedDaemonProcess.cs

```diff
@@ -66,18 +74,55 @@ internal sealed class MessageExtractor : HashCodeMessageExtractor
     {
         public MessageExtractor(int maxNumberOfShards)
             : base(maxNumberOfShards)
-        { }
+        {
+        }
+
+        public override string? EntityId(object message) => (message as ShardingEnvelope)?.EntityId;
+        public override object? EntityMessage(object message) => (message as ShardingEnvelope)?.Message;
 
-        public override string EntityId(object message) => (message as ShardingEnvelope)?.EntityId;
-        public override object EntityMessage(object message) => (message as ShardingEnvelope)?.Message;
-        public override string ShardId(string entityId, object messageHint = null)
+        public override string ShardId(string entityId, object? messageHint = null)
         {
             return entityId;
         }
     }
 
     /// <summary>
-    /// <para>This extension runs a pre set number of actors in a cluster.</para>
+    /// Used to support push-based communication with the <see cref="ShardedDaemonProcess"/> - uses
+    /// round-robin messaging to distribute messages to the actors in the cluster.
+    /// </summary>
+    /// <remarks>
+    /// NOTE: does not use a traditional <see cref="Router"/> to distribute messages because we want to use the
+    /// <see cref="ShardRegion"/> mechanism in order to piggyback off of its reliability and message buffering
+    /// features, which <see cref="Router"/> and <see cref="ActorSelection"/> do not support.
+    /// </remarks>
+    internal sealed class DaemonMessageRouter : UntypedActor
+    {
+        private readonly string[] _entityIds;
+        private readonly IActorRef _shardingRef;
+        private int _index = 0;
+
+        public DaemonMessageRouter(string[] entityIds, IActorRef shardingRef)
+        {
+            _entityIds = entityIds;
+            // validate that we have at least 1 entityId
+            if (_entityIds.Length == 0)
+                throw new ArgumentException("At least one entityId must be provided", nameof(entityIds));
+            _shardingRef = shardingRef;
+        }
+
+        protected override void OnReceive(object message)
+        {
+            var nextId = _entityIds[_index % _entityIds.Length];
+            
+            // have to remember to always allow the sharding envelope to be forwarded
+            _shardingRef.Forward(new ShardingEnvelope(nextId, message));
```

Comment:
Forward the message and wrap it inside a `ShardingEnvelope` so the underlying Akka.Cluster.Sharding infrastructure can safely deliver it.